### PR TITLE
[DXAA-553] Create Quickstart Fetch Utility

### PR DESCRIPTION
**Pull request by wdaimee**
### Changes

This PR introduces two shared utilities that fetch and cache quickstart specification data from the CDN. These utilities will be consumed by the future onboarding tools (`auth0_onboarding`, `auth0_get_quickstart_guide`, `auth0_save_credentials_to_file`). None of which call these utilities as of yet (will be addressed in future PRs).

The following functionalities were added:
1. Added `src/utils/fetch.ts` - a generic `fetchWithOptions` wrapper with configurable timeout (default 5s), opt-in retries, retry delay, and external abort signal support. Returns raw Response; 4xx is returned without retry, 5xx and network errors are retried.

2. Added `src/utils/quickstarts.ts` - a `fetchQuickstartSpec(framework)` function that resolves the CDN's two-step versioned URL structure (production.json -> versioned definition), strips unused fields (`download`, `steps`, `nextSteps`, `technologyLable`, etc.), and caches specs in-memory with 1h TTL and stale-on-error fallback.

3. Updated `src/utils/types.ts` - added `QuickstartSpec` interface with fields consumed by the MCP tools.

4. Added unit and integration tests for both utilities.

### References

https://auth0team.atlassian.net/browse/DXAA-553

### Testing

Neither utility is called by any tool handlers yet - they will be utilized when the onboarding tools are implemented. In the meantime, functionality is verified through unit tests.

Unit tests have been added to verify:

`fetchWithOptions`:
- successful fetch response is returned
- 4xx returned without retry
- 5xx not retried by default
- 5xx on subsequent retries, returns cached response (stale) if all retires fail
- 5xx retried, returns success if retry succeeds
- Network error retried, returns success if retry is successful
- Network error throws when all retries are exhausted unsuccessfully
- Custom retries, timeouts, and delay options respected

`fetchQuickstartSpec`:
- Fetches, caches, and returns the stripped spec
- Two-step URL resolution (production.json fetched -> version definition)
- Framework to filename mapping
- Unknown framework name returns null with no fetch
- Cache respected and hit within TTL (skips network call)
- stale-on-error returns cached spec when CDN fails after TTL
- Network error triggers retry

### Checklist

- [x] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [x] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
- [x] All existing and new tests complete without errors

### CODE REVIEW

**Comment by kushalshit27** [comment]
Moving it to draft. Please make it review-ready once tools are added.

**Comment by wdaimee** [comment]
Hi @kushalshit27, we are merging this branch into another branch specific for a new tool integration. This PR does not merge into master. The `feat/DXAA-get-quickstart-guide-tool-main` branch will be the branch/PR that we want your team to review. 

I am going to move this PR back to `Ready for review` as our team will review it and I will merge it to our working branch.

File: src/utils/quickstarts.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  const quickstartReleaseResponse = await fetchWithOptions(QUICKSTART_RELEASE_URL, FETCH_OPTIONS);
+  if (!quickstartReleaseResponse.ok) {
+    throw new Error(`Failed to fetch latest.json: ${quickstartReleaseResponse.status}`);
+  }
+
+  const { current: version } =
+    (await quickstartReleaseResponse.json()) as QuickstartReleaseResponse;
+
+  const fileName = FRAMEWORK_FILENAMES[framework];
+  const url = `${CDN_BASE}/versions/${version}/assets/definitions/${fileName}`;
```
Comment (GilbertLS) [inline src/utils/quickstarts.ts]: **Nit: might want to validate `version` before interpolating it into a URL**

The `version` comes straight from the CDN response and gets dropped into:
```typescript
const url = `${CDN_BASE}/versions/${version}/assets/definitions/${fileName}`;
```

If someone ever manages to mess with that CDN response (unlikely but not impossible), they could stick in `../../whatever` and point the fetch somewhere unexpected.

A quick check fixes it:
```typescript
if (!/^\d+\.\d+\.\d+$/.test(version)) {
  throw new Error(`Invalid version format from CDN: ${version}`);
}
```

Super low risk since it's our own CDN over HTTPS, but it's basically free defense-in-depth.

File: src/utils/quickstarts.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+const cache = new Map<string, CacheEntry>();
+
+class QuickstartCDNNotFoundError extends Error {
+  constructor(url: string) {
+    super(`Quickstart CDN resource not found: ${url}`);
+    this.name = 'QuickstartCDNNotFoundError';
+  }
+}
+
+const stripUnusedFields = (raw: any): QuickstartSpec => ({
```
Comment (GilbertLS) [inline src/utils/quickstarts.ts]: Right now `stripUnusedFields` takes `any` and just grabs properties off it. If the CDN response ever changes shape (field gets renamed, goes missing, whatever), you'll end up with a `QuickstartSpec` full of `undefined` values and no error thrown.

Should we add a joi or zod schema?

File: src/utils/quickstarts.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+interface QuickstartReleaseResponse {
+  name: string;
+  scheme: string;
+  current: string;
+  fallback: string;
+}
+
+const cache = new Map<string, CacheEntry>();
```
Comment (GilbertLS) [inline src/utils/quickstarts.ts:54]: **Minor: thundering herd on cache miss**

If a bunch of concurrent requests come in for the same framework while the cache is empty (or just expired), each one independently fires off two CDN fetches. They all succeed, all write to the cache, and you end up with N*2 network calls instead of 2.

Not a big deal today with only 4 frameworks, and it being self-hosted

```typescript
const inflight = new Map<string, Promise<QuickstartSpec | null>>();

// In fetchQuickstartSpec:
const existing = inflight.get(key);
if (existing) return existing;

const promise = fetchAndCache(key);
inflight.set(key, promise);
try { return await promise; } finally { inflight.delete(key); }
```

Non-blocking, just something to keep in mind.

File: test/utils/quickstarts.test.ts [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+      }),
+      mockDefinition('react')
+    );
+    await fetchQuickstartSpec('react');
+    expect(latestCalls).toBe(1);
+
+    await fetchQuickstartSpec('react');
+    expect(latestCalls).toBe(1);
+  });
+});
```
Comment (GilbertLS) [inline test/utils/quickstarts.test.ts:396]: **A few edge cases I didn't see covered**

- If `callBackPath` is missing from the CDN response, `stripUnusedFields` returns `callbackPath: undefined`, and downstream string concatenation would produce something like `'http://localhost:3000undefined'` — a malformed OAuth URL that silently passes through.
- If `production.json` returns valid JSON but is missing the `current` field, `version` is `undefined` and the URL becomes `.../versions/undefined/assets/...` — it'd fetch a nonsense path instead of throwing.
- Framework strings with mixed case like `'React'` or `'REACT'` — `.toLowerCase()` should handle this fine but the tests only ever pass lowercase inputs, so there's no verification.
- Two different frameworks fetched concurrently when the cache is empty — each one independently fires off CDN requests. Not really a test gap per se, but there's no concurrency test documenting the behavior, and it'll matter once this is called from concurrent MCP tool handlers.

File: src/utils/quickstarts.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+  placeholders: raw.placeholders,
+  environment: raw.environment,
+  llmPromptUrl: raw.llmPromptUrl,
+  envSnippet: raw.envSnippet,
+});
+
+const fetchFromCDN = async (framework: string): Promise<QuickstartSpec> => {
+  const quickstartReleaseResponse = await fetchWithOptions(QUICKSTART_RELEASE_URL, FETCH_OPTIONS);
+  if (!quickstartReleaseResponse.ok) {
+    throw new Error(`Failed to fetch latest.json: ${quickstartReleaseResponse.status}`);
```
Comment (GilbertLS) [inline src/utils/quickstarts.ts]: Failed to fetch latest.json

File: src/utils/quickstarts.ts [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+  constructor(url: string) {
+    super(`Quickstart CDN resource not found: ${url}`);
+    this.name = 'QuickstartCDNNotFoundError';
+  }
+}
+
+const stripUnusedFields = (raw: any): QuickstartSpec => ({
+  appType: raw.appType,
+  defaultAppOrigin: raw.defaultAppOrigin,
+  callbackPath: raw.callBackPath,
```
Comment (GilbertLS) [inline src/utils/quickstarts.ts]: callBackPath -> callbackPath

File: src/utils/types.ts [hunk 7]
```diff
@@ -75,9 +75,19 @@ export interface DefaultAppOrigin {
...
   logoutPath: string;
+  llmPromptUrl: string;
+  envSnippet: {
+    fileName: string;
+    requiredKeys: string[];
+    secretKeys: string[];
+  };
+  placeholders: Record<string, string>;
+  inputs: Record<string, string>;
+  environment: Record<string, any>;
```
Comment (GilbertLS) [inline src/utils/types.ts]: environment only has strings

```
  "environment": {
    "gitBranch": "quickstart/login",
    "gitRemote": "https://github.com/auth0-samples/auth0-react-samples",
    "environment": "remote",
    "downloadUrl": "https://github.com/auth0-samples/auth0-react-samples/archive/refs/heads/quickstart/login.zip",
    "filePathPrefix": "https://github.com/auth0-samples/auth0-react-samples/tree/quickstart/login/"
  }
```

**Review by GilbertLS — CHANGES_REQUESTED** [review]

File: src/utils/quickstarts.ts (hunk 2 above)
Comment (wdaimee) [reply src/utils/quickstarts.ts]: Using zod schema validation to parse the quickstart spec

**Review by GilbertLS — APPROVED** [review]
Looks good
